proposals/p3763.md

```diff
@@ -0,0 +1,858 @@
+
+Implicit conversions are performed as necessary to initialize the parameters of
+the function in the `impl` from the parameters of the thunk, and to initialize
+the return value of the function from the result of the call.
+
+It is an error if a thunk is needed to wrap a function declaration with a `var`
+parameter, because otherwise a copy would always be performed when initializing
+the parameter.
+
+> **Note:** These rules do not cover the case where the function declaration in
+> the `interface` or `impl` has implicit generic bindings. That case is
+> considered to be out of scope for this proposal.
+
+### `virtual` functions
+
+For a `virtual` function, the same approach is taken as for `impl` functions: a
+thunk is generated that differs from the declaration in the base class by
+replacing the type of `self` with the derived class, unless the function chosen
+for a class can be used directly. When a virtual function is used directly in a
+base class and not overridden in the derived class, it is also used directly in
+the derived class, even though its declared `self` parameter does not have a
+matching type.
+
+```carbon
+base class B {
+  // No thunk used.
+  virtual fn F[addr self: B*]();
+  virtual fn G[addr self: B*]();
+}
+base class C {
+  extend B;
+  // No thunk used: `self` has expected type `C*`.
+  impl fn F[addr self: C*]();
+  // Uses a thunk due to unexpected `self` type.
+  impl fn G[addr self: B*]();
+}
+class D {
+  // No thunk for `F`, because no thunk was used in `C`.
+  // Uses thunk for `F`, because thunk was used in `C`.
+  extend C;
+}
+```
+
+Note that this supports covariant return types automatically, as well as any
+other case where the return value from the derived class function can be
+implicitly converted to the base class function's return type. However, an
+`impl fn` doesn't introduce a new name lookup result, so the return type of a
+call expression is always that of the `virtual fn`, which means this feature is
+not useful.
+
+> **Future work:** It might be useful to allow a declaration to both implement
+> an existing virtual function and introduce a new one. This would allow
+> introducing functions with covariant return types that work as expected. This
+> could be achieved with syntax such as:
+>
+> ```carbon
+> base class A {
+>   virtual fn Clone[self: Self]() -> A*;
+> }
+> base class B {
+>   virtual impl fn Clone[self: Self]() -> B*;
+> }
+> ```
+>
+> Here, a call to `b->Clone()` would find `B.Clone` rather than `A.Clone`, and
+> so would have return type `B*`. The downside is that the vtable for `B` would
+> have two `Clone` slots, for `A.Clone` and `B.Clone`, whereas a covariant
+> return in C++ would only need a single vtable slot to express the same thing.
+
+It is an error for a class with a custom value representation to declare or
+implement a virtual function that passes `self` by value.
+
+### `extern` declarations
```

Comment:
Just wanted to make sure that @jonmeow had a chance to look over the latest version of this and it makes sense to him as well.